require(script.BombGame.GameModule)
require(script.HandlePurchases.PurchaseHandler)
require(script.HandlePurchases.InGamePurchases)
require(script.Inventory.InventoryEvents)
require(script.ProximityPromptHandler)
local GameTableAssigner = require(script.BombGame.GameTableAssigner)
local GlobalAnnouncer = require(script.MessageService.GlobalAnnouncer)
local ServerDataHandler = require(script.ServerData.ServerDataHandler)
local ServerPlayerEvents = require(script.ServerPlayerEvents)
local ServerSignalsInitializer = require(script.ServerSignalsInitializer)

local function ServerInit()
	ServerDataHandler.Initialize()
	ServerPlayerEvents.Handle()
	ServerSignalsInitializer.HandleEvents()
	GameTableAssigner:Initialize()
	GlobalAnnouncer.Initialize()
end

ServerInit()
